ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -73,29 +70,124 @@ import           Ouroboros.Consensus.Util.STM (blockUntilChanged)
 
--- | An action representing an update to the LoE fragment, that determines which
--- blocks can be selected in the ChainDB. With Ouroboros Genesis, this is
--- implemented via the GDD governor, see 'updateLoEFragGenesis'.
+-- | A never ending computation that evaluates the GDD rule whenever
+-- the STM action @getTrigger@ yields a different result, writing the LoE
+-- fragment to @varLoEFrag@, and then triggering ChainSel to reprocess all
+-- blocks that had previously been postponed by the LoE.
 --
--- The callback is applied to the current chain and the current ledger state,
--- and yields the new LoE fragment, which should be anchored in the immutable
--- tip.
-data UpdateLoEFrag m blk = UpdateLoEFrag {
-    updateLoEFrag ::
-         AnchoredFragment (Header blk)
-      -> ExtLedgerState blk
-      -> m (AnchoredFragment (Header blk))
-  }
-  deriving stock (Generic)
-  deriving anyclass (NoThunks)
+-- Evaluating the GDD rule might cause peers to be disconnected if they have
+-- sparser chains than the best chain.
+--
+-- The LoE fragment is the fragment anchored at the immutable tip and ending at
+-- the LoE tip.
+--
+-- @getHandles@ is the callback to get the handles that allow to disconnect
+-- from peers.
+--
+runGDDGovernor ::
+  ( Monoid a,
+    Eq a,
+    HasHardForkHistory blk,
+    IOLike m,
+    LedgerSupportsProtocol blk,
+    Ord peer
+  ) =>
+  TopLevelConfig blk ->
+  Tracer m (TraceGDDEvent peer blk) ->
+  STM m (Map peer (ChainSyncClientHandle m blk)) ->
+  StrictTVar m (AnchoredFragment (Header blk)) ->
+  ChainDB m blk ->
+  STM m a ->
+  m Void
+runGDDGovernor cfg tracer getHandles varLoEFrag chainDb getTrigger =
+  spin mempty
+  where
+    spin oldTrigger = do
+      (newTrigger, curChain, curLedger) <- atomically $ do
+        (_, newTrigger) <- blockUntilChanged id oldTrigger getTrigger
+        curChain <- ChainDB.getCurrentChain chainDb
+        curLedger <- ChainDB.getCurrentLedger chainDb
+        pure (newTrigger, curChain, curLedger)
+      loeFrag <- evaluateGDD cfg tracer getHandles curChain curLedger
+      oldLoEFrag <- atomically $ swapTVar varLoEFrag loeFrag
+      -- The chain selection only depends on the LoE tip, so there
+      -- is no point in retriggering it if the LoE tip hasn't changed.
+      when (AF.headHash oldLoEFrag /= AF.headHash loeFrag) $
+        triggerChainSelectionAsync chainDb
+      spin newTrigger
 
--- | A dummy version of the LoE that sets the LoE fragment to the current
--- selection. This can be seen as emulating Praos behavior.
-updateLoEFragUnconditional ::
-  MonadSTM m =>
-  UpdateLoEFrag m blk
-updateLoEFragUnconditional =
-  UpdateLoEFrag $ \ curChain _ -> pure curChain
+-- | Disconnect peers that lose density comparisons and recompute the LoE fragment.
+--
+-- Disconnecting peers causes candidate fragments to be removed, which causes
+-- the GDD governor to reevaluate GDD over and over until no more peers are
+-- disconnected.
```

Comment:
I think so. The GDD governor is retriggered on every modification to the list of candidate fragments. In particular, it will retrigger when candidate fragments are removed as an effect of disconnecting peers.

Reply:
> The GDD governor is retriggered on every modification to the list of candidate fragments.

Not on *every* modification, right? Citing from #1085 

> Lastly, we avoid triggering chain selection on every downloaded header. We do this by retriggering chain selection only when GDD changes the tip of the LoE fragment.

---

> In particular, it will retrigger when candidate fragments are removed as an effect of disconnecting peers.

:+1:

Reply:
> Not on every modification, right? Citing from #1085

Note that the citation is about chain selection, while the claims I'm making here are about the GDD governor. Do you think these are not accurate in some way?

Reply:
Ah you are right, I confused GDD and chain selection :+1: